File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int mul = 1;
    
        int n = nums.size();
        vector<int>ans(n, 0);
        // for(int i = 0; i < n; i++){
        //     for(int j = 0; j < n; j++){
        //         if(i == j) continue;
        //         mul *= nums[j];
        //     }
        //     ans[i] = mul;
        //     mul = 1;
        // }

        vector<int>left(n, 0);
        vector<int>right(n, 0);
        left[0] = 1;
        right[n-1] = 1;
        for(int i = 1; i < n; i++){
            mul = nums[i-1]*mul;
            left[i] = mul;
        }
        mul = 1;
        for(int i = n-2; i >= 0; i--){
            mul = nums[i+1] * mul;
            right[i] = mul;
        }
        
        for(int i = 0; i < n; i++){
            if(i == 0){
                ans[i] = right[i];
            }
            if(i == n-1){
                ans[i] = left[n-1];
            }
            ans[i] = left[i]*right[i];
        }

        return ans;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left = []
        right = []
        ans = []

        left.append(1)
        right.append(1)

        n = len(nums)

        pdct = 1

        for i in range(1, n):
            pdct = pdct * nums[i-1]
            left.append(pdct)
        
        pdct = 1

        for j in range(n-2, -1, -1):
            pdct = pdct * nums[j+1]
            right.append(pdct)

        right.reverse()

        for i in range(n):
            ans.append(left[i] * right[i])
        
        return ans